rpmlint/checks/SharedLibraryPolicyCheck.py

```diff
@@ -0,0 +1,174 @@
+        # already in BinaryCheck._check_shared_library
+        # Search for shared libraries in this package
+        libs = set()
+        libs_needed = set()
+        libs_to_dir = {}
+        dirs = set()
+        reqlibs = set()
+        pkg_requires = {x.name.split('(')[0] for x in pkg.requires}
+
+        for filename, pkgfile in pkg.files.items():
+            path = Path(filename)
+            if '.so.' in filename or filename.endswith('.so'):
+                if stat.S_ISREG(pkg.files[filename].mode) and pkgfile.magic.startswith('ELF '):
+                    readelf_parser = ReadelfParser(pkgfile.path, filename)
+                    failed_reason = readelf_parser.parsing_failed_reason()
+                    if failed_reason:
+                        self.output.add_info('E', pkg, 'readelf-failed', filename, failed_reason)
+                        return
+                    dyn_section = readelf_parser.dynamic_section_info
+                    libs_needed = libs_needed.union(dyn_section.needed)
+                    if dyn_section.soname:
+                        lib_dir = str(path.parent)
+                        libs.add(dyn_section.soname)
+                        libs_to_dir[dyn_section.soname] = lib_dir
+                        dirs.add(lib_dir)
+                    if dyn_section.soname in pkg_requires:
+                        # But not if the library is used by the pkg itself
+                        # This avoids program packages with their own
+                        # private lib
+                        # FIXME: we'd need to check if somebody else links
+                        # to this lib
+                        reqlibs.add(dyn_section.soname)
+
+        # If this is a program package (all libs it provides are
+        # required by itself), bail out
+        if not pkg.name.startswith('lib') and not libs.difference(reqlibs):
+            return
+
+        std_lib_package = False
+        pkg_ends_with_digit = pkg.name[-1].isdigit()
+        if pkg.name.startswith('lib') and pkg_ends_with_digit:
+            std_lib_package = True
+
+        # ignore libs in a versioned non_std_dir
+        if std_lib_package:
+            for lib in libs.copy():
+                lib_dir = libs_to_dir[lib]
+                for lib_part in lib_dir.split('/'):
+                    if not lib_part:
+                        continue
+                    if lib_part[-1].isdigit() and not lib_part.endswith('lib64'):
+                        libs.remove(lib)
+                        break
+
+        # Check for non-versioned libs in a std lib package
+        if std_lib_package:
+            for lib in libs.copy():
+                if (not (lib[-1].isdigit() or
+                         self.re_soname_strongly_versioned.search(lib))):
+                    self.output.add_info('W', pkg, 'shlib-unversioned-lib', lib)
+
+        if not pkg.name.startswith('lib') or pkg.name.endswith('-lang'):
```

Comment:
this return is weird can’t we match the regexp above and just either proceed or not.

Reply:
isn't this test too late ?

Reply:
We want to catch `shlib-unversioned-lib` also for a package not starting with `lib` if I'm correct.

Reply:
but there can't be slpp package without starting by lib*